lib/Barriers/PreDefined.pm

```diff
@@ -0,0 +1,317 @@
+   New_barrier_3 (labeled as 68) : max( round(barrier_3/m) * m, new_barrier_4 + m )
+   New_barrier_4 (labeled as 57) : max( round(barrier_4/m) * m, new_barrier_5 + m )
+   New_barrier_5 (labeled as 50) : rounded_central_spot
+   New_barrier_6 (labeled as 43) : min( round(barrier_6/m) * m, new_barrier_5 - m )
+   New_barrier_7 (labeled as 32) : min( round(barrier_7/m) * m, new_barrier_6 - m )
+   New_barrier_8 (labeled as 22) : min( round(barrier_8/m) * m, new_barrier_7 - m )
+   New_barrier_9 (labeled as 5)  : min( round(barrier_9/m) * m, new_barrier_8 - m )
+
+6) Apply the barriers for each contract types as defined in config file:
+   Example of config file:
+   - Single_barrier_european_option: [95, 78, 68, 57, 50, 43, 32, 22, 5]
+   - Single_barrier_american_option: [95, 78, 68, 57, 43, 32, 22, 5]
+   - Double_barrier_european_option: [68, 95, 57, 78, 50, 68, 43, 57, 32, 50, 22, 43, 5, 32],
+   - Double_barrier_american_option: [32, 68, 22, 78, 5, 95]
+
+=cut
+
+=head1 INPUT PARAMETERS
+
+=head2 config
+
+A configuration hashref that contains the selected barrier level for a contract type
+
+=head2 contract_type
+
+The contract type.
+
+=head2 duration
+
+The contract duration in seconds
+
+=head2 central_spot
+
+The spot at the start of the contract
+
+=head2 display_decimal
+
+The number of the display decimal point. Example 2 mean 0.01
+
+=head2 method
+
+The method for the barrier calculation, method_1 or method_2
+
+=cut
+
+=head2 _contract_barrier_levels
+
+A set of barrier level that intended to obtain for a contract type
+
+For example:
+   - Single_barrier_european_option: [95, 78, 68, 57, 50, 43, 32, 22, 5]
+   - Single_barrier_american_option: [95, 78, 68, 57, 43, 32, 22, 5]
+   - Double_barrier_european_option: [68, 95, 57, 78, 50, 68, 43, 57, 32, 50, 22, 43, 5, 32],
+   - Double_barrier_american_option: [32, 68, 22, 78, 5, 95]
+
+The barrier level 78 is 28 * min_barrier_interval from the central spot, while 22 is -28 * min_barrier_interval from the central spot.
+
+=cut
+
+has _contract_barrier_levels => (
+    is => 'rw',
+);
+
+has calculate_method_1 => (
+    is         => 'rw',
+    lazy_build => 1,
+);
+
+has calculate_method_2 => (
+    is         => 'rw',
+    lazy_build => 1,
+);
+
+=head2 BUILD
+
+We unwrap the config to map the barrier_level correspond to the contract type
+
+=cut
+
+sub BUILD {
+    my $self   = shift;
+    my $args   = shift;
+    my $config = $args->{config};
+    my $contract_barrier_levels;
+    for my $set (@$config) {
+        for my $type (@{$set->{types}}) {
+            $contract_barrier_levels->{$type} = $set->{barrier_level};
+        }
+    }
+
+    $self->_contract_barrier_levels($contract_barrier_levels);
+    return;
+}
+
+=head1 METHODS
+
+=cut
+
+=head2 calculate_available_barriers
+
+A function to calculate available barriers for a contract type
+Input_parameters: $contract_type, $duration, $central_spot, $display_decimal, $method
+
+=cut
+
+sub calculate_available_barriers {
```

Comment:
@basiliscos , I am not sure which one is better, but since this is not that complicated , maybe it is better not to make this as class, but just make it as a static method without creating an object.
- Just my 2 cents -